src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1616,15 +1616,28 @@ MasterAggregateExpression(Aggref *originalAggregate,
 		Expr *directarg;
 		foreach_ptr(directarg, originalAggregate->aggdirectargs)
 		{
-			if (!IsA(directarg, Const) && !IsA(directarg, Param))
+			/*
+			 * Need to overwrite Var nodes so that they point to the related
+			 * column of the result set returned for the worker aggregation.
+			 *
+			 * We also need to do the same for the Var's within complex
+			 * expressions, so we use pull_var_clause_default here.
+			 */
+			List *directArgVars = pull_var_clause_default((Node *) directarg);
+			if (list_length(directArgVars) > 0)
 			{
-				Var *var = makeVar(masterTableId, walkerContext->columnId,
-								   exprType((Node *) directarg),
-								   exprTypmod((Node *) directarg),
-								   exprCollation((Node *) directarg),
-								   columnLevelsUp);
-				aggregate->aggdirectargs = lappend(aggregate->aggdirectargs, var);
-				walkerContext->columnId++;
+				Var *directArgVar = NULL;
+				foreach_ptr(directArgVar, directArgVars)
+				{
+					Var *var = makeVar(masterTableId, walkerContext->columnId,
+									   exprType((Node *) directArgVar),
+									   exprTypmod((Node *) directArgVar),
+									   exprCollation((Node *) directArgVar),
+									   columnLevelsUp);
+					walkerContext->columnId++;
+
+					aggregate->aggdirectargs = lappend(aggregate->aggdirectargs, var);
+				}
```

Comment:
reconsider: Is it the correct approach to add each Var individually to the targetlist ?

Reply:
> reconsider: Is it the correct approach to add each Var individually to the targetlist ?

No. This is because, if we only collect Vars here, then we would ignore lots of other node types, as in `OpExpr(Const(2) * Var(a))`.



> I think one question is what is supposed to happen to something like `sum(Coalesce(somecolumn,3))` when there are 0 results.

Let me think about that little more

Reply:
> I think one question is what is supposed to happen to something like `sum(Coalesce(somecolumn,3))` when there are 0 results.

In my opinion; whenever we see a `Var` node within the expression, we should replace that node with a simple `Var` that refers to the related column of the worker agg result.